src/crypto/hpke/mod.rs

```diff
@@ -0,0 +1,236 @@
+                epoch: 0,
+            }
+        }
+
+        #[must_use]
+        pub fn seal(
+            &mut self,
+            key_id: KeyIdentifier,
+            match_key: XorReplicated,
+        ) -> MatchKeyEncryption {
+            let aad =
+                AssociatedData::new(key_id, self.epoch, &Self::HELPER_ORIGIN, &Self::SITE_ORIGIN)
+                    .unwrap();
+            let info = aad.to_bytes();
+            let mut plaintext = [0_u8; 16];
+            match_key.serialize(&mut plaintext).unwrap();
+            let pk_r = self.registry.public_key(self.epoch, key_id);
+
+            let (encap_key, tag) =
+                single_shot_seal_in_place_detached::<IpaAead, super::IpaKdf, super::IpaKem, _>(
+                    &OpModeS::Base,
+                    pk_r,
+                    &info,
+                    &mut plaintext,
+                    &[],
+                    &mut self.rng,
+                )
+                .unwrap();
+
+            MatchKeyEncryption {
+                encap_key: <[u8; 32]>::from(encap_key.to_bytes()),
+                ct: Box::new(plaintext),
+                tag: <[u8; 16]>::from(tag.to_bytes()),
+            }
+        }
+
+        pub fn open(
+            &self,
+            key_id: KeyIdentifier,
+            mut enc: MatchKeyEncryption,
+        ) -> Result<XorReplicated, BottomError> {
+            let aad =
+                AssociatedData::new(key_id, self.epoch, &Self::HELPER_ORIGIN, &Self::SITE_ORIGIN)
+                    .unwrap();
+            open_in_place(
+                &self.registry,
+                &enc.encap_key,
+                enc.ct.as_mut(),
+                &enc.tag,
+                &aad,
+            )?;
+
+            Ok(XorReplicated::deserialize(enc.ct.as_ref())?)
+        }
+
+        pub fn advance_epoch(&mut self) {
+            self.epoch += 1;
+        }
+    }
+
+    /// Make sure we obey the spec
+    #[test]
+    fn ipa_info_serialize() {
+        let aad =
+            AssociatedData::new(255, 32767, &HelperOrigin("foo"), &SiteOrigin("bar")).unwrap();
+        assert_eq!(
+            b"private-attributionfoo\0bar\0\xff\x7f\xff",
+            aad.to_bytes().as_ref()
+        );
+    }
+
+    #[test]
+    fn decrypt_happy_case() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(1, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+
+        let enc = suite.seal(0, match_key);
+        let r = suite.open(0, enc).unwrap();
+
+        assert_eq!(match_key, r);
+    }
+
+    #[test]
+    fn decrypt_wrong_aad() {
+        let rng = StdRng::from_seed([1_u8; 32]);
+        let mut suite = EncryptionSuite::new(1, rng);
+        let match_key = XorReplicated::new(u64::MAX, u64::MAX / 2);
+        let enc = suite.seal(0, match_key);
+        suite.advance_epoch();
+
+        let _ = suite.open(0, enc).unwrap_err();
+    }
+
+    #[test]
+    fn decrypt_wrong_key() {
```

Comment:
Also, when we don't have the identified key.

Reply:
that seems to be validated at compile time unless I misunderstood the question here

Reply:
How do you mean?  If I pass in a key identifier that you don't know about (say 100 in this test where you only have 0..10), I think the code will crash.

Reply:
ah that's right. That goes to the comment that we shouldn't crash if we don't have the key